### ISSUE TITLE
Initial design decisions

### ISSUE THREAD

**Issue by lquenti** [BasedApeYachtClub/codash#3]
We should initially decide a few things. Just add if you think anything is missing:

## Code style

No strongly held opinions on that.

## Docstring standard

I think there is just doxygen in the C space. Sadly, it is not opinionated at all...

## Testing Library

I think Google Test is a good choice, even if a more lightweight solution would be enough. Also, I want to learn it.

## Build system

`make` seems sufficient imo.

**Comment by lquenti** [comment]
## C-Standard Compatibiltity

I don't wanna cast all `void *` just to do some basic arithmetic on them, thus I'd propose that we stick with at least that GCC extension

**Comment by meipp** [comment]
# Parameter order and name conventions
Additionally to code style we should use the same function signatures for similar functions. This includes:
- whether to use `xs`, `ys`, `zs` for arrays or `src`, `dst`
- `len`, `size` or `xs_len`, `x_size`

**Comment by meipp** [comment]
> ## Docstring standard
> 
> I think there is just doxygen in the C space. Sadly, it is not opinionated at all...

I propose we use doxygen javadoc style and we only document the header files.

```
/**
 * @brief Checks if all elements of an array satisfy the given predicate
 * 
 * @param xs The array
 * @param len The number of elements in `xs`
 * @param size The size of each element in `xs`
 * @param predicate The predicate
 * @return true if `predicate` is true for all elements in `xs`
 * @return false if `predicate` is not true match all elements in `xs` (i.e. is false for some)
 */
void *find_first(void *xs, size_t len, size_t size, bool (*predicate)(void *));
```

Note that doxygen supports `\`` formatting.

**Comment by meipp** [comment]
> ## Testing Library
> 
> I think Google Test is a good choice, even if a more lightweight solution would be enough. Also, I want to learn it.

#5 Sorry about that...

**Comment by meipp** [comment]
> ## Build system
> 
> `make` seems sufficient imo.

I agree. We should use `make` for building, running the tests, and generating the docs.
